### ISSUE TITLE
Inquiry About Guidepup and VoiceOver Automation

### ISSUE THREAD

**Issue by zhangjiurui** [guidepup/guidepup#99]
Hi Craig,

I hope you are doing well. My name is Raymond, and I’m a Quality Engineer working on accessibility testing. At my company, we’re currently exploring automation frameworks that can integrate macOS VoiceOver with our existing stack — which includes Playwright, React, TypeScript, and Yarn.

This week, I began a proof of concept using several GitHub frameworks, and I came across Guidepup, which looks very promising for VoiceOver automation. As I’m not a Front-End developer, I’m still getting familiar with the details of VoiceOver and its integration.

If possible, could you please share a contact method (email or public discussion forum) where I could ask a few questions and learn more about Guidepup?

Here are a few things I’m curious about:

1. Reporting: Is there a way to generate Playwright-style reports that show the status of VoiceOver tests, including failure points and detailed logs or error messages?
2. Headless Execution: Can tests be run in a headless mode — i.e., without VoiceOver being visibly active on the Mac?
3. If no headless available, is it possible to speed up the execution faster?
4. Tutorials: Are there any step-by-step video tutorials available, similar to Playwright’s documentation, that walk through Guidepup usage?I’d really appreciate any guidance or resources you could share.

Best regards,
Raymond

**Comment by cmorten** [comment]
Hi @zhangjiurui 👋 

Answers in order:

1. If you integrate with a test framework such as Playwright then you can generate test reports same as you would with any other tests. Guidepup itself isn’t a testing framework so doesn’t have its own reporters, but pairs well with testing frameworks such as Playwright, Puppeteer, Jest, Vitest which do. See https://github.com/guidepup/aria-at-tests as an example repo using Playwright with Guidepup to perform a comprehensive set of tests using VoiceOver and NVDA which generates a report with screen reader output and recordings available at https://aria-at-tests.vercel.app/
2. Most “real” screen readers are targeted at real users and don’t support headless modes, so there’s no way currently to run VoiceOver etc headlessly. Best alternatives are to consider the likes of Axe for static analysis tests, inspecting the accessibility tree yourself using something like [Playwright accessibility snapshot](https://playwright.dev/docs/api/class-accessibility#deprecated) or [other methods that I’ve written about previously](https://medium.com/john-lewis-software-engineering/automating-a11y-testing-part-3-testing-library-and-beyond-f4cd590c7395), or you can try out the Guidepup Virtual Screen Reader https://github.com/guidepup/virtual-screen-reader which is a package aimed at simulating screen readers based on W3C specs for headless testing.
3. Being a framework that drives VoiceOver speed is limited by VoiceOver itself which unfortunately isn’t fast… issues/PRs welcome if have any ideas/suggestions. Users have reported that using the `capture: "initial"` option when starting VO brings the speed to a more acceptable level (see https://www.guidepup.dev/docs/api/class-command-options#command-options-capture)
4. There is a docs website at https://www.guidepup.dev/ that is inspired by Playwright’s docs site. Each Guidepup repo also has a set of relevant examples in an examples directory, eg see https://github.com/guidepup/guidepup/tree/main/examples for this repos examples.

**Comment by zhangjiurui** [comment]
Thanks for the reply.  

- for the CI, will explore the options for the MacOS/Windins Agents at Gitlab.

- For the password frozen issue, the password (Testing@123) was entered as expected. but the log script after it was not run. The below is the original html code. As you mentioned, I should try the Playwright to do the password typing instead of the 'voiceOver.type()'

`<input type="password" name="password" required="" aria-invalid="true" aria-describedby="passwordError" autocomplete="current-password" id="passwordInput" class="sc-e3cf0185-0 sc-43bd9ed8-1 sc-32dd1043-2 ejtAYo cncIbX jStsIh" value="">`
